### PR TITLE
[SPARK-16448] RemoveAliasOnlyProject should not remove alias with metadata

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

`Alias` with metadata is not a no-op and we should not strip it in `RemoveAliasOnlyProject` rule.
This PR also did some improvement for this rule:
1. extend the semantic of `alias-only`. Now we allow the project list to be partially aliased.
2. add unit test for this rule.
## How was this patch tested?

new `RemoveAliasOnlyProjectSuite`

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @viirya @yhuai @liancheng

**Comment by viirya** [comment]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -165,36 +165,48 @@ object PushProjectThroughSample extends Rule[LogicalPlan] {
  * but can also benefit other operators.
  */
 object RemoveAliasOnlyProject extends Rule[LogicalPlan] {
-  // Check if projectList in the Project node has the same attribute names and ordering
-  // as its child node.
+  /**
+   * Returns true if the project list is semantically same with child output, after strip alias on
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Nit: "... same with ..." => "... same as ..."

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -165,36 +165,48 @@ object PushProjectThroughSample extends Rule[LogicalPlan] {
...
+      stripAliasOnAttribute(projectList).zip(childOutput).forall {
+        case (a: Attribute, o) if a semanticEquals o => true
+        case _ => false
       }
     }
   }
 
+  private def stripAliasOnAttribute(projectList: Seq[NamedExpression]) = {
+    projectList.map {
+      // Alias with metadata can not be striped, or the metadata will be lost.
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Nit: striped => stripped

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -165,36 +165,48 @@ object PushProjectThroughSample extends Rule[LogicalPlan] {
...
+      }
+      val attrMap = AttributeMap(attributesToReplace)
       plan.transformAllExpressions {
         case a: Attribute if attrMap.contains(a) => attrMap(a)
       }.transform {
-        case op: Project if op.eq(p) => op.child
+        case plan: Project if plan eq proj => plan.child
       }
     }.getOrElse(plan)
   }
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Can we use a `plan.transform` to implement this rule?

**Comment by liancheng** [comment]
LGTM except for some minor comments.

**Comment by liancheng** [comment]
Merging to master.

**Pull request closed without merging**